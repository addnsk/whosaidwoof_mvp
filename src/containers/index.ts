export * from "./Header";
export * from "./Theme";
export * from "./SelectList";
