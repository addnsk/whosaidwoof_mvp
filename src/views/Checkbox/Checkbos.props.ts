export interface ICheckboxProps
  extends React.DetailedHTMLProps<
    React.InputHTMLAttributes<HTMLInputElement>,
    HTMLInputElement
  > {
  caption?: () => JSX.Element;
}
