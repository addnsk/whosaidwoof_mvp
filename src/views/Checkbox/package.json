{
  "name": "checkbox-view",
  "version": "0.1.0",
  "private": true,
  "main": "Checkbox.tsx"
}
