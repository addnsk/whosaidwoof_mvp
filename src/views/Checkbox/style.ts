import styled from "styled-components";

export const Label = styled.label`
  display: flex;
  align-items: center;
  width: fit-content;

  cursor: pointer;

  outline: none;
  user-select: none;

  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
`;

export const Caption = styled.span`
  margin-left: 1.5vh;
  max-width: 40vw;
`;

export const VisbilityHidden = styled.input`
  width: 1px;
  height: 1px;
  clip: rect(0 0 0 0);
  overflow: hidden;
  margin: 0;
  padding: 0;
  appearance: none;

  outline: none;
`;

export const Division = styled.div`
  width: 20px;
  height: 20px;
  border: 1px solid ${({ theme }) => theme.colors.light};
  border-radius: 5px;

  position: relative;

  ${VisbilityHidden}:checked + &:after {
    content: "";
    width: 12px;
    height: 12px;
    border-radius: 3px;
    background-color: ${({ theme }) => theme.colors.light};

    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
  }

  ${VisbilityHidden}:focus + & {
    box-shadow: 0 0 0 1px ${({ theme }) => theme.colors.accentLight};
  }
`;
