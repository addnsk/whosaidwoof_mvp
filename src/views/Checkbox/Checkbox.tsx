import * as React from "react";
import { ICheckboxProps } from "./Checkbos.props";
import { Label, VisbilityHidden, Division, Caption } from "./style";

export const Checkbox: React.FC<ICheckboxProps> = (props): JSX.Element => {
  const { caption, checked, disabled, onChange } = props;
  return (
    <Label>
      <VisbilityHidden
        type="checkbox"
        checked={checked}
        disabled={disabled}
        onChange={onChange}
      />
      <Division />
      <Caption>{caption && caption()}</Caption>
    </Label>
  );
};
