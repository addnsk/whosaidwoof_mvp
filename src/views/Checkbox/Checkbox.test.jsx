import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { Theme } from "containers";
import { Checkbox } from "./Checkbox";
import userEvent from "@testing-library/user-event";

const onChange = jest.fn();

describe("Checkbox component", () => {
  it("Checkbox  renders with full props", () => {
    render(
      <Theme>
        <Checkbox
          checked={false}
          caption={() => <span>Hello</span>}
          onChange={onChange}
        />
      </Theme>
    );
    expect(screen.getByRole("checkbox")).toBeInTheDocument();
    expect(screen.getByRole("checkbox")).toHaveStyle({
      appearance: "none",
    });

    expect(screen.getByText(/hello/i)).toBeInTheDocument();
    expect(screen.getByText(/hello/i).tagName).toEqual("SPAN");

    expect(screen.getByRole("checkbox")).toMatchSnapshot();
  });

  it("Checkbox renders with some default props and children", () => {
    render(
      <Theme>
        <Checkbox caption={() => <span>Hello</span>}>Goodbye</Checkbox>
      </Theme>
    );
    expect(screen.getByRole("checkbox")).toBeInTheDocument();

    expect(screen.getByText(/hello/i)).toBeInTheDocument();
    expect(screen.queryByText(/goodbye/i)).toBeNull();

    expect(screen.getByRole("checkbox")).toMatchSnapshot();
  });

  it("Checkbox renders without props", () => {
    render(
      <Theme>
        <Checkbox />
      </Theme>
    );
    expect(screen.getByRole("checkbox")).toBeInTheDocument();
    expect(screen.getByRole("checkbox").firstChild).toBeNull();

    expect(screen.getByRole("checkbox")).toMatchSnapshot();
  });

  it("Checkbox responds to onClick", () => {
    render(
      <Theme>
        <Checkbox
          checked={false}
          caption={() => <span>Hello</span>}
          onChange={onChange}
        />
      </Theme>
    );

    userEvent.click(screen.getByRole("checkbox"));
    userEvent.click(screen.getByText(/hello/i));

    expect(onChange).toHaveBeenCalledTimes(2);
    expect(screen.getByRole("checkbox")).not.toBeChecked();
  });
});
