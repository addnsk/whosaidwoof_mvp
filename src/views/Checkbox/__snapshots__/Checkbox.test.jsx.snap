// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component Checkbox  renders with full props 1`] = `
<input
  class="sc-jSgvzq gUpjID"
  type="checkbox"
/>
`;

exports[`Checkbox component Checkbox renders with some default props and children 1`] = `
<input
  class="sc-jSgvzq gUpjID"
  type="checkbox"
/>
`;

exports[`Checkbox component Checkbox renders without props 1`] = `
<input
  class="sc-jSgvzq gUpjID"
  type="checkbox"
/>
`;
